iteration documentation and limitations based on unit tests first draft (#189)
AleoHQ/welcome: documentation/leo/03_language.md

```

#### Mapping
The mapping struct allows the programmer to apply updates to a program mapping data structure by calling one of the
following functions.

#### get
A get command, e.g. `current_value = Mapping::get(counter, addr);`
Gets the value stored at `addr` in `counter` and stores the result in `current_value`
If the value at `addr` does not exist, then the program will fail to execute.

#### get_or_init
A get command that initializes the mapping in case of failure, e.g.
`let current_value: u64 = Mapping::get_or_init(counter, addr, 0u64);`
Gets the value stored at `addr` in `counter` and stores the result in `current_value`.
If the key is not present, `0u64` is stored in `counter` and stored in `current_value`.

#### set
A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
Sets the `addr` entry as `current_value + 1u64` in `counter`.

## For Loops

> For Loops are declared as `for <loop variable> in <lower bound>..<upper bound>`.  Currently, the

Reviewer: For consistency with the existing docs, let's change <> to {}, so `for {variable} in {lower-bound}..{upper-bound}`. A good example of this to reference is the [mapping type](https://developer.aleo.org/leo/language#mapping).

loop variable supports unsigned integer types u8, u16, u32, or u64. The lower bound must be

> less than the upper bound.

Reviewer: After this sentence, add "Nested loops are supported." and remove the example of a nested loop below.

```leo
for <loop variable> in <lower bound>..<upper bound> {

>     // Loop body

Reviewer: Same thing here change <> to {}

}
```

### Nested Loops

> You can nest loops within loops. For example:

Reviewer: I don't think you need to provide an example here. Just say nested loops are supported.

loop iterates over 0, 1, 2).

In the loop body, you can use the loop variable `i`. In every iteration, `i` takes the value of the current iteration.
In this example, `i` takes the values 0, 1, 2 in sequence.

## Loop Behavior

### Nested Loops

You can nest loops within loops. For example:

```leo
for i: u32 in 0u32..2u32 {
    for j: u32 in 0u32..2u32 {
        // inner loop body
    }
}
```

In this case, for each value of `i`, the inner loop runs fully, iterating over its entire range of values.

### Non-Literal Loop Bounds

Leo does not currently support non-literal loop bounds. For example, the following code will fail:

```leo
for i:u64 in 0u64..amount {
    // loop body
}
```

The value of `amount` is not known at compile time, so the loop cannot be unrolled and compiled.

### Return Statements in Loops

Leo does not currently support return statements within loops. A function in Leo can only have one return statement, and
it must be the last statement in the function. For example, the following code will fail:

```leo
for i: u32 in 0u32..9u32 {
    return false;
}
```

The `return` statement is inside the loop, which is not allowed in Leo.

> ## Loop Errors

Author: Good point! Would an alternate implementation for non-literal loop bounds be asserting that the bound isn't too large?

### Loop Bounds

Leo checks loop bounds at compile time. If the start bound is greater than the end bound, Leo will throw a compile
error. For example, the following code will fail:

```leo
for i: u32 in 9u32..0u32 {
    // loop body
}
```

The start bound 9 is greater than the end bound 0, which is not valid.

### Large Loop Bounds

Leo does not support loop bounds larger than the maximum representable value of the loop variable's type. For example,
the following code will fail:

```leo
for i:u64 in 0u64..1000000000000000000000000000000000000000000000000000000000000000000000000000000000000u64 {
    // loop body
}
```

The end bound is too large to be represented as a `u64`.

## Notes

1. Leo doesn't currently support the `break` and `continue` statements found in other languages such as
   Rust. These statements allow for more complex control flow within loops, but are not currently
   available in Leo.

> 

Reviewer 2: make these more specific: say exactly what the minimum and maximum loop requirements are or remove these docs

A get command that initializes the mapping in case of failure, e.g.
`let current_value: u64 = Mapping::get_or_init(counter, addr, 0u64);`
Gets the value stored at `addr` in `counter` and stores the result in `current_value`.
If the key is not present, `0u64` is stored in `counter` and stored in `current_value`.

#### set
A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
Sets the `addr` entry as `current_value + 1u64` in `counter`.

## For Loops

For Loops are declared as `for {loop variable} in {lower bound}..{upper bound}`.  Currently, the

> loop variable supports unsigned integer types u8, u16, u32, or u64. The lower bound must be

Reviewer: @moAlobaidi, this should change to support up to u32. Just got confirmation from the Leo team. For context, internally we support up to u128 but externally we shouldn't mention this.

> For Loops are declared as `for {loop variable} in {lower bound}..{upper bound}`.  Currently, the
> loop variable supports unsigned integer types u8, u16, u32, or u64. The lower bound must be

Reviewer 2: ```suggestion
For Loops are declared as `for {loop variable} in {lower bound}..{upper bound}`.  Typical Leo loop variable types include `u8`, `u16`, and `u32`. The lower bound must be
```

Reviewer 2: What is a `{loop variable}` it is not defined in the documentation.

Reviewer 2: That description still does not help the programmer write a loop @moAlobaidi.
Loops in Leo require a variable name and a variable type.
`{loop variable}` should be defined as `{name: type}` with an example.

Author: Okay updated.

less than the upper bound. Nested loops are supported.

```leo
for {loop variable} in {lower bound}..{upper bound} {
    // Loop body
}
```

> 

Reviewer 2: You should add a code snippet of a loop that actually compiles.
People will want to copy and paste code from the docs into their program to play around with the feature.

```

#### Mapping

The mapping struct allows the programmer to apply updates to a program mapping data structure by calling one of the
following functions.

#### get

A get command, e.g. `current_value = Mapping::get(counter, addr);`
Gets the value stored at `addr` in `counter` and stores the result in `current_value`
If the value at `addr` does not exist, then the program will fail to execute.

#### get_or_init

A get command that initializes the mapping in case of failure, e.g.
`let current_value: u64 = Mapping::get_or_init(counter, addr, 0u64);`
Gets the value stored at `addr` in `counter` and stores the result in `current_value`.
If the key is not present, `0u64` is stored in `counter` and stored in `current_value`.

#### set

A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
Sets the `addr` entry as `current_value + 1u64` in `counter`.

## For Loops

> For Loops are declared as `for {variable: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer 

Reviewer 2: ```suggestion
For Loops are declared as `for {variable: type} in {lower bound}..{upper bound}`. Unsigned integer 
```
The lower and upper bounds are literals - their type is specified with the value `1u8`, `1337u16`
There is no colon required @moAlobaidi

types `u8`, `u16`, and `u32` are recommended for loop variable types. The lower bound must be
less than the upper bound. Nested loops are supported.

### Example

```leo
  let count: u32 = 0u32;

  for i: u32 in 0u32..5u32 {
    count += 1u32;
  }

>   return count;

Reviewer 2: ```suggestion
  return count; // returns 5u32
```

```

#### Mapping

The mapping struct allows the programmer to apply updates to a program mapping data structure by calling one of the
following functions.

#### get

A get command, e.g. `current_value = Mapping::get(counter, addr);`
Gets the value stored at `addr` in `counter` and stores the result in `current_value`
If the value at `addr` does not exist, then the program will fail to execute.

#### get_or_init

A get command that initializes the mapping in case of failure, e.g.
`let current_value: u64 = Mapping::get_or_init(counter, addr, 0u64);`
Gets the value stored at `addr` in `counter` and stores the result in `current_value`.
If the key is not present, `0u64` is stored in `counter` and stored in `current_value`.

#### set

A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
Sets the `addr` entry as `current_value + 1u64` in `counter`.

## For Loops

> For Loops are declared as `for {variable_name: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer 

Reviewer 2: ```suggestion
For Loops are declared as `for {variable: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer 
```
Remove mixed usage of snake_case and text in description

> types of up to
> u32 are recommended for iterator variables. The lower bound must be

Reviewer 2: ```suggestion
types `u8`, `u16`, and `u32` are recommended for loop variable types. The lower bound must be
```
* State explicit types.
* Add backticks to types so they show up as `code` - this is consistent with the rest of `u8` mentions throughout the documentation.
* Remove usage of "iterator" name which has not been mentioned before in documentation
* Specify that we are describing `type` of. the iterator variable.

less than the upper bound. Nested loops are supported.

### Example

```leo
  let count: u32 = 0u32;

>   for i:u32 in 0u32..5u32 {

Reviewer 2: ```suggestion
  for i: u32 in 0u32..5u32 {
```
